### DISCUSSION TITLE
How do you make map for game (pacman) in bevy

### DISCUSSION THREAD

**Discussion by Abubakar1122331** [bevyengine/bevy#7065] [Q&A]
Hi! There,
Im Trying to make pacman clone.
Here is my txt file that im trying to use
[map.txt](https://github.com/bevyengine/bevy/files/10324776/map.txt).

`pub fn create_simple_map(mut commands: Commands, asset_server: Res<AssetServer>) {
    let file = File::open("assets/map.txt").expect("No map file found");
    
    for (y, line) in BufReader::new(file).lines().enumerate() {
        if let Ok(line) = line {
            for (a, char) in line.chars().enumerate() {
                char as usize;

                if char == '#' {
                    commands.spawn(SpriteBundle{
                        texture: asset_server.load("dot.png"),
                        transform: Transform{
                            scale: Vec3::new( 1.0, 1.0, 1.0),
                            ..default()
                        },
                        ..default()
                    });
                }
            }
        }
    }
}`

But when i start the game it's like this.
<img width="1442" alt="image" src="https://user-images.githubusercontent.com/110607297/210072522-9052a9fe-a188-4683-8be1-b55dddb01c04.png">

Look at the center of image.
there is the whole map. (if you can see it!)

**Comment by rparrett** [comment]
You will need to spawn the sprites making up the map with different `translation` (positions) according to their locations in the map file.

```rust
const GRID_SIZE: f32 = 30.;

pub fn create_simple_map(mut commands: Commands, asset_server: ResMut<AssetServer>) {
    let file = File::open("assets/map.txt").expect("No map file found");

    for (row, line) in BufReader::new(file).lines().flatten().enumerate() {
        for (col, char) in line.chars().enumerate() {
            if char == '#' {
                commands.spawn(SpriteBundle {
                    texture: asset_server.load("dot.png"),
                    transform: Transform {
                        translation: Vec3::new(col as f32 * GRID_SIZE, row as f32 * GRID_SIZE, 0.),
                        scale: Vec3::new(1.0, 1.0, 1.0),
                        ..default()
                    },
                    ..default()
                });
            }
        }
    }
}
```

**Reply by Abubakar1122331** [reply]
Thanks Bro.
but how do you calculate GRID_SIZE.
😁👌

**Reply by Abubakar1122331** [reply]
<img width="1443" alt="image" src="https://user-images.githubusercontent.com/110607297/210094707-1fb74a71-6704-4c67-b47e-ebd91e8aca16.png">
And it's Stretched, Why....

**Reply by rparrett** [reply]
`GRID_SIZE` would be an artistic choice. Use a value that works with your assets.

As for the stretching, looking at your `map.txt`, the columns are separated by commas and spaces, and these are being counted for the `col` variable when enumerating. (side note: `col` and `row` were swapped in my original answer. I fixed that in an edit.)

Here are a couple strategies you could use to deal with that:
```rust
// Filter "non-map characters" prior to enumeration
for (col, char) in line.chars().filter(|c| *c == '#' || *c == '.').enumerate() {

// Use "split" to deal with the column separators instead of iterating over characters
for (col, char) in line.split(", ").enumerate() {
```

**Reply by Abubakar1122331** [reply]
How to change the translation of the whole map.

**Reply by Abubakar1122331** [reply]
And it's not spawning all of them correctly.

**Reply by rparrett** [reply]
You can define an offset and add it to the translation when spawning the map:
```rust
const MAP_OFFSET: Vec3 = Vec3::new(-150., -150., 0.);

// ...

translation: MAP_OFFSET + Vec3::new(col as f32 * GRID_SIZE, row as f32 * GRID_SIZE, 0.),
```

> And it's not spawning all of them correctly.

That's not a lot of information to go on!

**Reply by Abubakar1122331** [reply]
Look the pattern is not right
<img width="1442" alt="image" src="https://user-images.githubusercontent.com/110607297/210110172-088444cc-e8d4-4812-a0a1-403215618bc6.png">

**Reply by rparrett** [reply]
It looks like`row` and `col` in the translation are mixed up like I mentioned above.

**Reply by Abubakar1122331** [reply]
So how can i fix it?
Any Idea?

**Reply by rparrett** [reply]
```diff
- Vec3::new(row as f32 * GRID_SIZE, col as f32 * GRID_SIZE, 0.)
+ Vec3::new(col as f32 * GRID_SIZE, row as f32 * GRID_SIZE, 0.)
```

**Reply by Abubakar1122331** [reply]
Oh now it works
Thanks,
i didn't setted it because it was rotating all the map but now it's OK👌

**Reply by Abubakar1122331** [reply]
How to rotate the whole map?

**Reply by rparrett** [reply]
If you just need some sort of 90 degree rotation or flip, you may be able to accomplish what you need by changing the order you are iterating through the map file with e.g. [rev](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.rev).

Otherwise, you can spawn the sprites as children of a single `SpatialBundle` and set the `rotation` field of that parent bundle's `Transform` to `Quat::from_rotation_z(angle_in_radians)`.

Here are some resources that might help with that:

https://bevy-cheatbook.github.io/features/parent-child.html
https://bevy-cheatbook.github.io/features/transforms.html

Good luck!

**Reply by Abubakar1122331** [reply]
I have made a spatial bundle and set the tilemap as it's children, but it's transform doesn't effect the map.

`{pub fn spawn_cell_points(
    mut commands: Commands,
    asset_server: Res<AssetServer>,
    mut windows: ResMut<Windows>,
    mut texture_atlases: ResMut<Assets<TextureAtlas>>,
) {

    
    let window = windows.get_primary_mut().unwrap();
    let (w, h) = (window.width(), window.height());
    let file = File::open("assets/map.txt").expect("No map file found");
    let MAP_OFFSET: Vec3 = Vec3::new(-w / 2. + 100.0 -( w / 4.3 ), h / 2. + 28., 0.);
    let MAP_SCALE: Vec3 =  Vec3::new(-1., -1.,-1.);
    let mut tiles = Vec::new();


    for (col, line) in BufReader::new(file).lines().flatten().enumerate() {
            for (row, char) in line.chars().enumerate() {

                let texture_handle = asset_server.load("dot.png");
                let texture_atlas = TextureAtlas::from_grid(texture_handle, Vec2::new(16.0, 16.0), 1, 1, None, None);
                let texture_atlas_handle = texture_atlases.add(texture_atlas);

                if char == '.' {
                    let tile = commands.spawn(SpriteSheetBundle {
                        texture_atlas: texture_atlas_handle,
                        transform: Transform {
                            translation: MAP_OFFSET + Vec3::new(row as f32 * GRID_SIZE * 1.6, -(col as f32) * GRID_SIZE * 3.2, 1.),
                            scale: MAP_SCALE - Vec3::new(1., 1., 3.0),
                            ..default()
                        },
                        ..default()
                    })
                    .insert(FuelCell { capacity: 10.0 })
                    .insert(Velocity { x: 0.0, y: 0.0})
                    .insert(Name::new("Food")).id();
                    tiles.push(tile);
                }

            }
        }
        commands
        .spawn(SpatialBundle{
            transform: Transform {
                translation: Vec3::new(-100., 0.0, 0.0),
                scale: Vec3::new(10.0, 1.0, 1.0),
                ..default()
            },
            visibility: Visibility {is_visible: true},
            ..default()
        })
        .insert(Name::new("Map"))
        .push_children(&tiles)
        .insert(TransformBundle::default());
}}`

**Reply by rparrett** [reply]
side note: If you use three backticks to surround your code, github won't mangle it. You can even specify a language for syntax highlighting.
~~~
```rust
code
```
~~~

It looks like the issue is that you are creating a command that spawns a `SpatialBundle` with a transform, but then immediately inserts (and overwrites) the transform with `.insert(TransformBundle::default())`.

Try removing that last insert.

There's also some other oddities that may or may not be contributing.
`scale: MAP_SCALE - Vec3::new(1., 1., 3.0)`

It's not clear to me what the intention is here, but scaling the z axis in 2d by anything other than `1.0` can only lead to strange behavior.

**Reply by Abubakar1122331** [reply]
Yep! that was scaling issue. I didn't thought that it can create error.😅

**Reply by Abubakar1122331** [reply]
Its Stretching, why?

**Reply by rparrett** [reply]
This looks suspicous
```rust
scale: Vec3::new(10.0, 1.0, 1.0),
```

**Reply by Abubakar1122331** [reply]
It's alr 

``` rust

commands
        .spawn(SpatialBundle{
            transform: Transform {
                translation: Vec3::new(1.0, 1.0, 0.0),
                scale: Vec3::new(1., 1., 1.0),
                ..default()
            },
            visibility: Visibility {is_visible: true},
            ..default()
        })
        .insert(Name::new("Wall"))
        .push_children(&walls)
        .insert(Velocity { x: 0.0, y: 0.0 })
        .insert(FuelCell { capacity: 10.0 });

**Reply by Abubakar1122331** [reply]
Ok, no problem I have fixed it 👍.
